File: 78.subsets.js
/*
 * @lc app=leetcode id=78 lang=javascript
 *
 * [78] Subsets
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const res = [];
    const subset = [];
    function helper(res, subset, pool){

        res.push([...subset]);
        for(let i=0; i<pool.length;i++){
            subset.push(pool[i]);
            helper(res, subset, pool.slice(i+1, pool.length));
            subset.pop();
        }
    }
        console.log(res);
    helper(res, subset, nums);
    return res;
    
};
subsets([1,2,3]);
// @lc code=end

File: 78.subsets.java
/*
 * @lc app=leetcode id=78 lang=java
 *
 * [78] Subsets
 */

// @lc code=start
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> subsets(int[] nums) {
        dfs(nums, 0, new ArrayList<>());
        return res;
        
    }
    void dfs(int[] nums, int start, List<Integer> path){
        //这里直接就上来放进去了
        //后面这段就可以不可以的
        res.add(new ArrayList<>(path));
        for(int i=start; i<nums.length; i++){
            //not added
            int cur = nums[i];
            path.add(cur);
            dfs(nums, i+1, path);
            path.remove(path.size()-1);
        }

    }
}
// @lc code=end

